### DISCUSSION TITLE
GD32 MCU support

### DISCUSSION THREAD

**Discussion by sakisdog** [iNavFlight/inav#8455] [Ideas]
The last 2 days i have been experimenting with a dead OMNIBUS F4 V3 flight controller. The STM32F405 chip was shorted between VDD and VSS so i ordered a STM replacement and a GigaDevice replacement (GD32F405RGT6).
The GD32 chip is pin-to-pin identical to STM32F405 with the only difference on 2 pins that the STM have connected 2 caps.
I soldered the GD32F405 on the dead flight controller and after confirming that the short is gone i power it up with the boot button pressed. And...surprise! The GD32 DFU device appeared in my device list.
I installed the WinUSB drivers through Zadig but Inav configurator could not "see" the device. I changed the PID and VID of usbDevices inside the inav configurator with the correct ones and finally i was able to flash the flight controller.
The flight controller seems to boot up normally with the blue led blinking fast many times during the startup and then blinking with a slow rate.
The problem is that i cannot see the controller in my device list. When i connect it with usb nothing happens. No plug-in sound from windows, no device in device manager. Nothing. It just sits there.
After searching a lot in chinese forums i concluded that it is a firmware problem. So now i'm trying to find what parts of code i should change for the usb to work.
Anyone have any clues?

**Comment by babitsyn** [comment]
I use GDAllInOneProrammer to flash GD32F103CBT6 installed in CC3D with inav 1.7.3, and it works well. I think it's good idea to return support fo F3 targets due to GigaDevices offers pin-to-pin compatible chip GD32F303CGT6 in LQFP-48 package with 256kB program memory and 768 kB data memory onboard operated at 120 MHz clock.  It would be cheap replacement for stm32F103 and will revive CC3D boards.

**Reply by sakisdog** [reply]
@babitsyn I'm glad to hear that! Usb works ok for you?

**Reply by babitsyn** [reply]
Yes, USB work good. Now I set sysclk  frequency at 108 MHz, and USB not recognized, probably due to bad divider for USB clock which set in firmware. This divider has only 2 options and didn't calculated, just sysclk compared to 72 MHz - equal or not. I suspect "sysclk not equal to 72 MHz" means USE_UNDERCLOCK option. May be you know what is underclock frequency for F1 target?

**Reply by sakisdog** [reply]
1. What files did you modify?
2. When you say "USB not recognized" you mean that windows display a message that usb is not recognized or that the device does not enumerate at all?

**Reply by babitsyn** [reply]
I modified \inav\lib\main\CMSIS\CM3\DeviceSupport\ST\STM32F10x\system_stm32f10x.c:

_#define SYSCLK_FREQ_72MHz  96000000
uint32_t SystemCoreClock = SYSCLK_FREQ_72MHz;   /*!< System Clock Frequency (Core Clock) */_

to change sysclk, but USB became not recognized device in Windows. It occurs because terrifying surprises in the code, like this: 

_#if defined(CJMCU) || defined(CC3D)
    // On CJMCU new revision boards (Late 2014) bit 15 of GPIOC->IDR is '1'.
    RCC_CFGR_PLLMUL = RCC_CFGR_PLLMULL9;
#else_

The PLL **multiplier is hardcoded** and set 9, so sysclk still 72 MHz, i.e. HSE*PLLMUL9 = 8 MHz * 9 = 72 MHz. And this hardcoded multipliers found in several files in lib and src directories.

**Reply by babitsyn** [reply]
Now I will create array "pllmult[]" which should be consist of uint32_t multipliers defined at stm32f10x.h like "#define  RCC_CFGR_PLLMULL15                 ((uint32_t)0x00340000)", remove all hardcoded multipliers and redefine it using simple expression: pllmult[((uint8_t) SystemCoreClock / HSE)] to set proper divider for USB. 
I think you need make same checks for GD32F405 definitions to set correct SystemCoreClock and USB clock. GD23F103 allows set USB clock dividers according to higher SystemCoreClock, for example: 1, 1.5, 2, and 2.5. Carefully read datasheet and user manual for your chip.

**Reply by sakisdog** [reply]
I already checked the SystemCoreClock and USB clock. It seems fine. I even hardcoded the values to be 100% sure but still no device showed up. The CoreClock of GD32F405 is the same as ST32F405 so i don't think i should make any changes.
I tested the chip (as i had it soldered on the flight controller) with the cdc_acm example project from the GD32 firmware library and it works fine. USB shows up and it enumarates normally. 
I check line-by-line the ST code in the inav project with the example project and it is the same. I'm missing something but i don't know what!

**Reply by sakisdog** [reply]
> I modified \inav\lib\main\CMSIS\CM3\DeviceSupport\ST\STM32F10x\system_stm32f10x.c:
> 
> _#define SYSCLK_FREQ_72MHz 96000000 uint32_t SystemCoreClock = SYSCLK_FREQ_72MHz; /*!< System Clock Frequency (Core Clock) */_
> 
> to change sysclk, but USB became not recognized device in Windows. It occurs because terrifying surprises in the code, like this:
> 
> _#if defined(CJMCU) || defined(CC3D) // On CJMCU new revision boards (Late 2014) bit 15 of GPIOC->IDR is '1'. RCC_CFGR_PLLMUL = RCC_CFGR_PLLMULL9; #else_
> 
> The PLL **multiplier is hardcoded** and set 9, so sysclk still 72 MHz, i.e. HSE*PLLMUL9 = 8 MHz * 9 = 72 MHz. And this hardcoded multipliers found in several files in lib and src directories.

Also in "hw_config.c" i found the following line that i believe is your problem. This code is in iNav 4.1. I don't know if it is present in previous versions.
`RCC_USBCLKConfig(RCC_Clocks.SYSCLK_Frequency == 72000000 ? RCC_USBCLKSource_PLLCLK_1Div5 : RCC_USBCLKSource_PLLCLK_Div1);`

**Reply by babitsyn** [reply]
This string **_RCC_USBCLKConfig(RCC_Clocks.SYSCLK_Frequency == 72000000 ? RCC_USBCLKSource_PLLCLK_1Div5 : RCC_USBCLKSource_PLLCLK_Div1);_**
I replaced with:
_**switch(RCC_Clocks.SYSCLK_Frequency){
		case 72000000:
			RCC_USBCLKConfig(RCC_USBCLKSource_PLLCLK_1Div5);
			break;
		case 48000000:
			RCC_USBCLKConfig(RCC_USBCLKSource_PLLCLK_Div1);
			break;
		case 96000000:
			RCC_USBCLKConfig(RCC_USBCLKSource_PLLCLK_Div2);
			break;
		case 120000000:
			RCC_USBCLKConfig(RCC_USBCLKSource_PLLCLK_2Div5);
			break;
		default:
			RCC_USBCLKConfig(RCC_USBCLKSource_PLLCLK_2Div5);
	}**_
And added definitions for multipliers from GD32 library.

**Reply by sakisdog** [reply]
Nice job! Did you solve the usb issue with this code?
I'm still trying to figure out the usb enumation issue. :(

**Reply by babitsyn** [reply]
I managed to run usb with a sysclk 120 MHz. I changed PLLMULL to 15 in normal mode and 9 for underclock mode, i.e. 120 MHz and 72 MHz respectively. But USB was not working due to USB prescaler set only 1 bit in stm32 library, instead of 2 bits in GD32 lib.  To run USB I modified **stm32f10x_rcc.h**:
     _#define RCC_USBCLKSource_PLLCLK_1Div5   ((uint8_t)0x00) // Clk 72 MHz
      #define RCC_USBCLKSource_PLLCLK_Div1    ((uint8_t)0x01) // Clk 48 MHz_
      **_#define RCC_USBCLKSource_PLLCLK_2Div5   ((uint8_t)0x02) // GD32 USB prescaler Clk 120 MHz
      #define RCC_USBCLKSource_PLLCLK_Div2    ((uint8_t)0x03) // GD32 USB prescaler Clk 96 MHz_**
Then modified **stm32f10x_rcc.c**:
_#ifndef STM32F10X_CL
 #define USBPRE_BitNumber          0x16
 #define CFGR_USBPRE_BB            (PERIPH_BB_BASE + (CFGR_OFFSET * 32) + (USBPRE_BitNumber * 4))_
**#define CFGR_USBPRE_BB1           (PERIPH_BB_BASE + (CFGR_OFFSET * 32) + ((USBPRE_BitNumber+1) * 4)) // Address of GD32 second bit of USB prescaler**
And add in same file:
void RCC_USBCLKConfig(uint32_t RCC_USBCLKSource)
{
  /* Check the parameters */
  assert_param(IS_RCC_USBCLK_SOURCE(RCC_USBCLKSource));

  *(__IO uint32_t *) CFGR_USBPRE_BB = RCC_USBCLKSource;
_**// Set second bit of USB prescaler
  *(__IO uint32_t *) CFGR_USBPRE_BB1 = RCC_USBCLKSource>>1;**_ 
}
So USB working now!

**Reply by babitsyn** [reply]
Just now checked USB functionality with cpu_underclock=ON, USB still OK.

**Reply by sakisdog** [reply]
Before you do this modifications, did you notice if usb was enumarated (even with errors) in windows? I mean, when you plugged the usb did you hear the "plug in" sound from windows or nothing happened?

**Reply by sakisdog** [reply]
> But USB was not working due to USB prescaler set only 1 bit in stm32 library, instead of 2 bits in GD32 lib

Also how did you find out about the 2 bits prescaler? What file did you check in GD32 lib? I'm asking because i want to check the F4 library and compare the files between GD32 and STM32. Maybe this is the change i have to doito be able to resolve my issues too.

**Reply by babitsyn** [reply]
> Before you do this modifications, did you notice if usb was enumarated (even with errors) in windows? I mean, when you plugged the usb did you hear the "plug in" sound from windows or nothing happened?

Yes, unknown USB device has appeared in device list.



> > But USB was not working due to USB prescaler set only 1 bit in stm32 library, instead of 2 bits in GD32 lib
> 
> Also how did you find out about the 2 bits prescaler? What file did you check in GD32 lib? I'm asking because i want to check the F4 library and compare the files between GD32 and STM32. Maybe this is the change i have to doito be able to resolve my issues too.

**gd32f10x_rcu.h** 
You should find this:

_/* USBD/USBFS prescaler select */
#define CFG0_USBPSC(regval)             (BITS(22,23) & ((uint32_t)(regval) << 22))
#define RCU_CKUSB_CKPLL_DIV1_5          CFG0_USBPSC(0)                      /*!< USBD/USBFS prescaler select CK_PLL/1.5 */
#define RCU_CKUSB_CKPLL_DIV1            CFG0_USBPSC(1)                      /*!< USBD/USBFS prescaler select CK_PLL/1 */
#define RCU_CKUSB_CKPLL_DIV2_5          CFG0_USBPSC(2)                      /*!< USBD/USBFS prescaler select CK_PLL/2.5 */
#define RCU_CKUSB_CKPLL_DIV2            CFG0_USBPSC(3)                      /*!< USBD/USBFS prescaler select CK_PLL/2 */_

As you see values 2 and 3 needs 2 bits.

**Reply by babitsyn** [reply]
Also I recommend flash firmware through USB-TTL converter connected to UART1 of stm32F4. And try connect INAV-configurator to this port and check VCP config at ports configuration menu.

**Reply by ujara** [reply]
I have soldered on an F3 board with a  GD32F303CCTX (it had a STM chip earlier). Any clues how to get it working with USB and 120Mhz? with INAV? I am trying to follow the thread above. I have modified the following files:

1. stm32f30x_rcc.h
2. stm32f30x_rcc.c
3. system_stm32f30x.c
4. hw_config.c
Were there any other changes required? I did try and look around for hard coded 72Mhz or 8 Mhz HSE cystal instance but seems i have covered it all. What I am not sure about are:
for GD32f303 can I use #ifndef STM32F30X_CL ? or is it something else?
Secondly for the changes in stm32f30x_rcc.h do I have to modify 

#define IS_RCC_USBCLK_SOURCE(SOURCE) (((SOURCE) == RCC_USBCLKSource_PLLCLK_1Div5) || \
                                      ((SOURCE) == RCC_USBCLKSource_PLLCLK_Div1))
To account for 120Mhz?       // Full speed

Last was the PLLMULL settings in system_stm32F30x.c where they as follows:
      // Full speed
        if (HSE_VALUE == 12000000) {
            RCC->CFGR |= (uint32_t)(RCC_CFGR_PLLSRC_PREDIV1 | RCC_CFGR_PLLXTPRE_PREDIV1 | RCC_CFGR_PLLMULL15);
        }
        else {
            RCC->CFGR |= (uint32_t)(RCC_CFGR_PLLSRC_PREDIV1 | RCC_CFGR_PLLXTPRE_PREDIV1 | RCC_CFGR_PLLMULL9);
        }
I guess I am not that great at coding and trying to guess some of it so your input is helpful..

**Reply by sakisdog** [reply]
I don't know about F3 chips but i had no luck with F4 and iNav. The only logical thing that i didn't try was to use the GigaDevice's Firmware Library with iNav but that requires that i have to import all the necessary files into iNav.
I compared all the offset addresses between STM32F4 and GD32F4 and i couldn't find a difference. I can not figure out why the HAL library of ST isn't compatible with GD.

**Reply by babitsyn** [reply]
You should try this sequence:
1. Press boot-key
2. Connect via usb-ttl adapter to UART1
3. Run GD All-in-one-Programmer
4. Erase whole chip 
5. Flash it.
6. Release boot-key. Run normally. 
Do not use Inav-configurator to flash GD, every time which I did it - GD32 never started correctly.

**Reply by ujara** [reply]
I do use the GD All-in-one-Programmer and followed what you had indicated but did not change the outcome.

**Reply by babitsyn** [reply]
> I have soldered on an F3 board with a GD32F303CCTX (it had a STM chip earlier). Any clues how to get it working with USB and 120Mhz? with INAV? I am trying to follow the thread above. I have modified the following files:
> 
>     1. stm32f30x_rcc.h
> 
>     2. stm32f30x_rcc.c
> 
>     3. system_stm32f30x.c
> 
>     4. hw_config.c
>        Were there any other changes required? I did try and look around for hard coded 72Mhz or 8 Mhz HSE cystal instance but seems i have covered it all. What I am not sure about are:
>        for GD32f303 can I use #ifndef STM32F30X_CL ? or is it something else?
>        Secondly for the changes in stm32f30x_rcc.h do I have to modify
> 
> 
> #define IS_RCC_USBCLK_SOURCE(SOURCE) (((SOURCE) == RCC_USBCLKSource_PLLCLK_1Div5) || ((SOURCE) == RCC_USBCLKSource_PLLCLK_Div1)) To account for 120Mhz? // Full speed
> 
> Last was the PLLMULL settings in system_stm32F30x.c where they as follows: // Full speed if (HSE_VALUE == 12000000) { RCC->CFGR |= (uint32_t)(RCC_CFGR_PLLSRC_PREDIV1 | RCC_CFGR_PLLXTPRE_PREDIV1 | RCC_CFGR_PLLMULL15); } else { RCC->CFGR |= (uint32_t)(RCC_CFGR_PLLSRC_PREDIV1 | RCC_CFGR_PLLXTPRE_PREDIV1 | RCC_CFGR_PLLMULL9); } I guess I am not that great at coding and trying to guess some of it so your input is helpful..

You did not set second bit of USB multiplier, please read previous messages. And check frequency of you crystal - 8 MHz or 12 MHz.
RCC_CFGR_PLLMULL15 and HSE=12 MHz will set cpuclock to 180 MHz, it's wrong.  For HSE=12 MHz PLLMUL should be 10.

**Reply by DaLLuZZ** [reply]
Maybe this will help you solve the problem with gigadevice f4 chips
https://forum.chibios.org/viewtopic.php?t=6008
https://forum.chibios.org/viewtopic.php?t=5758

**Comment by DzikuVx** [comment]
Inav will not bring back F1 or F3 support

**Reply by MicroMontis** [reply]
F1 and F3 are the past. Let's not look to the past. But what about the supporting of the GD32 F405 from GigaDevices ? Would you consider looking into this? Or should we just stick with STM?
